### ISSUE TITLE
`PwBaseWorkChain`: Add outputs of final `PwCalculation` when failed

### ISSUE THREAD

**Issue by mbercx** [aiidateam/aiida-quantumespresso#923]
The `PwBaseWorkChain` currently doesn't add the outputs of the wrapped `PwCalculation` when it is `Finished` but with a non-zero exit code:

```console
(aiida-super) aiida@prnmarvelsrv3:~/envs/aiida-super$ verdi process show 1988400
Property     Value
-----------  -------------------------------------------------------------------------------
type         PwBaseWorkChain
state        Finished [300] The calculation failed with an unidentified unrecoverable error.
pk           1988400
uuid         11d52d76-8362-4034-8a81-c9327f255014
label
description
ctime        2023-04-22 20:04:58.246858+00:00
mtime        2023-04-22 20:11:27.622940+00:00

Inputs                PK       Type
--------------------  -------  -------------
pw
    pseudos
        Ta            238      UpfData
        Fe            1897473  UpfData
        B             241      UpfData
    code              1946942  InstalledCode
    structure         1946799  StructureData
    parameters        1988391  Dict
    parallelization   1988393  Dict
clean_workdir         1988394  Bool
kpoints_distance      1988395  Float
kpoints_force_parity  1988397  Bool
max_iterations        1988398  Int

Called                             PK  Type
----------------------------  -------  ----------------------------
create_kpoints_from_distance  1988405  create_kpoints_from_distance
iteration_01                  1988423  PwCalculation

Log messages
----------------------------------------------
There are 5 log messages for this calculation
Run 'verdi process report 1988400' to see them
```

This makes obtaining results from the work chain less consistent, and e.g. someone that wants to restart from the last `parent_folder` will suddenly have to go a level deeper to obtain this. Querying for data that is _always_ there (e.g. walltime) is also more complicated. 

We should also add the outputs of the final `PwCalculation` to those of the `PwBaseWorkChain`. Let's also double check this is the case for other `BaseRestartWorkChain`s.

**Comment by sphuber** [comment]
This is probably the case for all `BaseRestartWorkChain`s. If a handler returns an exit code, indicating that the workchain should be aborted, it will do so straight away and so the `results` method is not called, which is what attaches the outputs. So unless an implementation manually calls the `results` method, the outputs will not be attached for a failing workchain. Looking at the implementation, it would probably be possible to simply call `self.results()` from within a process handler before it returns the exit code, which should have the desired effect. This probably won't work if it is done in the last iteration. In that case, a check in the `results` method will return `ERROR_MAXIMUM_ITERATIONS_EXCEEDED` instead. We should probably fix this in `aiida-core` where the output attaching code is factored out of `results` to something like `_attach_outputs` and just call that. This will allow implementations to call this method from handlers before exiting if they still want to forward the outputs. I will open a PR

**Comment by mbercx** [comment]
Yeah, fair point, was also considering this and came back to see you had the same conclusion. ^^ Let's keep this issue open until we fix it in `aiida-core`.

**Comment by sphuber** [comment]
> We'd also have to adapt our work chains if I understand correctly.

Yes, with the changes I propose we would still need to make use of it in the handlers. There might be ways of making this automatic. There are various methods on the base `Process` class that trigger when the state changes. We could override these to automatically call `_attach_outputs`. We already do this in the `BaseRestartWorkChain` where we override `on_terminated` to clean the working directories of the calcjobs. We could think to also use this method to attach the outputs of the last calcjob. The downside is that it will be then difficult to again disable this behavior for a specific implementation. They could re-override `on_terminated` but if they want to keep the cleaning they would have to copy this, which is not so nice from a maintenance perspective.

**Comment by mbercx** [comment]
Hmm, ok, thank for the explanation @sphuber! I'd say it's not urgent, so let's take some time to consider and write down the design.
